### PR TITLE
Add new API to allow inserter items to be prioritised

### PR DESCRIPTION
**Pull request by scruffian**
## What?

Adds a new API to the block list settings to allow parent blocks to define a priority for blocks in the inserter for a given inner blocks.

## Why?

In https://github.com/WordPress/gutenberg/pull/48752 we added a private API to allow blocks to be prioritised. This was so we could force `Page Link` and `Custom Link` to show up as the first two results in the inserter options for the Navigation block's list view editing mode.

However, this was always envisaged as a short term way to allow the wider feature to land in WP 6.2. This PR circles back (and is an alternative to https://github.com/WordPress/gutenberg/pull/49959) with an alternative approach which looks at this problem more holistically by assuming the requirement to prioritise blocks will likely be useful to other block's in future.

## How?

Adds a new `inserterPriority` property to inner blocks which allows the consumer to pass an array of blocks and/or variations which should be prioritised in the Inserter.

It also adds a concrete example, by way of alternating the Nav block to utilise this new property to cause the `Page Link` and `Custom Link` blocks to show up first in the Inserter.

#### Automated Testing

Run

```
npm run test:e2e packages/e2e-tests/specs/editor/plugins/inner-blocks-prioritized-inserter-blocks.test.js
```

#### Manual Testing

- Alter the `INSERTER_PRIORITY` constant in `constants.js` (see changeset) to be:

```js
export const INSERTER_PRIORITY = [
	'core/spacer',
	'core/navigation-link/page',
	'core/navigation-link',
];
```
- Rebuild with `npm run dev`
- New Post
- Add Nav block
- Open inspector controls
- Open List View tab in inspector controls
- Click inserter
- See `core/spacer` block as first item followed by the other two block variations.

## Screenshots or screencast 

Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

### CODE REVIEW

**Comment by getdave** [comment]
@scruffian As requested (away from Github) I reworked this to use a new API on inner blocks called `inserterPriority`. 

Why? Because I thought if we reuse `allowedBlocks` then any implementations that currently use `allowedBlocks` will have their inserter items unexpectedly automatically sorted.

Note that the `inserterPriority` items are automatically filtered by the `allowedBlocks` so if it contains blocks/variations that are not supported then they will be removed.

I guess we might need to do the same for the variations though, in case someone includes a variation of a block that isn't in the allowed list.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (draganescu) [inline packages/block-editor/src/components/inserter/search-results.js:111]: I see we had `maxBlockTypes` and `maxBlockTypesToShow` both existed in code. Why did we remove it here?

File: packages/block-editor/src/components/inserter/search-results.js
Comment (draganescu) [inline packages/block-editor/src/components/inserter/search-results.js]:
```suggestion
				// Sort block items according to `parentInserterPriority`.
```

File: packages/block-editor/src/components/inserter/search-results.js
Comment (getdave) [reply packages/block-editor/src/components/inserter/search-results.js:111]: I think the eslint rule detected a redundant prop. This wasn't due to our change here but was redundant from a previous commit. `maxBlockTypes` is not used within the `useMemo` hook so it's not needed. Perhaps it was used in the past.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (ajlende) [inline packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]: There's no reason for this to be memoized.

And unless `useMemo` handles dependencies differently from `useLayoutEffect`, the previous memo does nothing either. If `allowedBlocks` is getting a new object reference each time, the `_allowedBlocks` will also update to the new reference.

Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]: This is confusing but I think it's correct.

The dependencies list of `useMemo` is an array of dependencies. In this case the dependencies are the strings contained _within_ the `inserterPriority` array, not the `inserterPriority` _array_ itself.

Without the memo, any non-stable `inserterPriority` array would trigger the layout effect because the reference would change.

With the memo, even if the array reference is unstable the memo will fix it to make it stable. This will avoid the layout effect being called unnecessarily.

The memo "works" because the 2nd argument of the `useMemo` call is the _array of strings_. As these strings will stay the same even if the `inserterPriority` array _reference_ changes, the `useMemo` never needs to recompute.

Originally in https://github.com/WordPress/gutenberg/pull/30311/

File: packages/block-editor/src/components/inserter/index.js
Comment (getdave) [inline packages/block-editor/src/components/inserter/index.js:427]: We can probably follow up to remove the need for the private wrapper once (and if!) this PR is merged.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (getdave) [inline packages/block-editor/src/components/inserter/search-results.js]: I think we _might_ need to compute this value inside a `useSelect` hook. Why? Because currently if the block list or `rootClientId` were to change then this component would not update. `useSelect` would re-run the selectors if dependencies change so we might need to refactor toward that.

**Comment by getdave** [comment]
As this represents a change to a key API, I'm pinging @youknowriad for review as well.

File: packages/block-editor/src/components/inserter/index.js
Comment (talldan) [inline packages/block-editor/src/components/inserter/index.js:433]: I suppose the other option is to keep this API, but instead support passing a custom appender to List View (instead of the showAppender prop).

It still doesn't solve the issue where the block itself ends up with different ordering, unless both the block and the list view instance use the same custom appender, so I think what's in this PR is better. It's define once and it works everywhere which means there's less room for errors.

**Review by talldan — COMMENTED** [review]
Thank you for exploring this. It may have taken longer, but it feels like the better option to me.

I didn't have a chance to do a full code review, but the code looks like what I'd expect.

I'm not sure if it could be made private, but it also feels like a really useful feature to me, so it could shipped stable. Tests and docs might be needed once there's an agreement on this.

**Comment by youknowriad** [comment]
If I'm not wrong @ntsekouras did some work in the past for internal ordering in the inserter. So he might have some feedback here.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (draganescu) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]: Great explanation! 👏🏻

File: packages/block-editor/src/components/inserter/search-results.js
Comment (draganescu) [inline packages/block-editor/src/components/inserter/search-results.js]: What does this comment mean?

File: packages/block-editor/src/components/inserter/search-results.js
Comment (draganescu) [reply packages/block-editor/src/components/inserter/search-results.js:111]: 🆗

File: packages/block-editor/src/components/inserter/index.js
Comment (draganescu) [reply packages/block-editor/src/components/inserter/index.js:433]: I think 

> instead support passing a custom appender to List View (instead of the showAppender prop).

is better than the prop, but why would we need to keep `orderInitialBlockItems`?

**Comment by draganescu** [comment]
This looks excellent IMO. it solves the block inserter being tailored to what the block accepts as inner blocks and to what the most common UX is.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (ntsekouras) [inline packages/block-editor/src/components/inserter/search-results.js]: Nits: No need to return an object here and also no need to prepend the `parent` in the var name IMO.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (ntsekouras) [inline packages/block-editor/src/components/inserter/search-results.js]: This functionality would be great to add in slash inserter too [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/autocompleters/block.js#L67).

File: packages/block-library/src/navigation/edit/inner-blocks.js
Comment (ntsekouras) [inline packages/block-library/src/navigation/edit/inner-blocks.js]: I think a different name more close to what it actually is, would be better. Maybe `prioritizedInserterBlocks` or something like that..

File: packages/block-editor/src/components/inserter/search-results.js
Comment (getdave) [reply packages/block-editor/src/components/inserter/search-results.js]: Haha. Sorry yes. I should have removed that. It was to remind me that `.sort()` will _mutate_ rather than return a new reference like things like `.filter()` would do.

File: packages/block-editor/src/components/inserter/index.js
Comment (getdave) [reply packages/block-editor/src/components/inserter/index.js:433]: Only if you wanted _specific_ instance X of an inserter to have a specific order.

The new implementation in this PR is a blanket application. Any block that uses the new prop on `InnerBlocks` no matter where it is with have items prioritised.

The previous implementation was more targetted.

**Review by ntsekouras — COMMENTED** [review]
Thanks for the PR @scruffian! I haven't tested thoroughly, but it looks good.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (getdave) [reply packages/block-editor/src/components/inserter/search-results.js]: Counter argument is expressed in this post under heading `Prefer returning objects with properties from the callback`

https://jsnajdr.wordpress.com/2021/01/22/some-best-practices-for-using-useselect-from-wordpress-data/#:~:text=%7D-,Prefer,-returning%20objects%20with

File: packages/block-editor/src/components/inserter/search-results.js
Comment (getdave) [reply packages/block-editor/src/components/inserter/search-results.js]: I'll take a look at this.

File: packages/block-library/src/navigation/edit/inner-blocks.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/inner-blocks.js]: Ok no problem.

Comment (getdave) [reply packages/block-library/src/navigation/edit/inner-blocks.js]: ✅

File: packages/block-editor/src/components/inserter/search-results.js
Comment (getdave) [reply packages/block-editor/src/components/inserter/search-results.js]: This would involve extracting a hook for the retrieval of `prioritisedInserterBlocks` and also moving the sorting function to a shared utils location.

Once done this should be simple enough. I can tackle tomorrow.

Comment (getdave) [reply packages/block-editor/src/components/inserter/search-results.js]: This would involve extracting a hook for the retrieval of `prioritisedInserterBlocks` and also moving the sorting function to a shared utils location.

Once done this should be simple enough. I can tackle tomorrow.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (ntsekouras) [reply packages/block-editor/src/components/inserter/search-results.js]: No strong opinion here :) . In our case the `prioritizedInserterBlocks` though is a `const` and doesn't change.

**Review by MaggieCabrera — APPROVED** [review]
This is testing nicely, the e2e tests were passing locally. Let's see if GH can handle those today...

File: packages/block-library/src/navigation/constants.js
Comment (MaggieCabrera) [inline packages/block-library/src/navigation/constants.js]: did we mean to put the spacer here or is this for testing?

Comment (getdave) [reply packages/block-library/src/navigation/constants.js]: Nice catch 😓

File: packages/block-editor/src/components/inserter/search-results.js
Comment (scruffian) [reply packages/block-editor/src/components/inserter/search-results.js]: > This functionality would be great to add in slash inserter too [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/autocompleters/block.js#L67).

Let's do this in a followup.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (ajlende) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]: The comment should explain it like that—the existing one isn't very clear when scanning over the code.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (ajlende) [inline packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]:
```suggestion
	// Memoize allowedBlocks and prioritisedInnerBlocks based on the contents
	// of the arrays. Implementors often pass a new array on every render,
	// and the contents of the arrays are just strings, so the entire array
	// can be passed as dependencies.
	
```

**Comment by ntsekouras** [comment]
Please before merging this, docs and some tests should be added, as is a new API.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (ntsekouras) [reply packages/block-editor/src/components/inserter/search-results.js]: We can, but in general this is a tiny change. I tried it before commenting. We just need to find a good place for the sorting function(ex utils or something). Also we don't even need a new hook, as is just a `select` call.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]: @ajlende Yeh you have a point there. I can raise a PR to augment the existing comments here if you don't fancy taking it on.

**Comment by getdave** [comment]
A good next step here would be to identify whether there are any tests covering the `allowedBlocks` feature? If not then we will need to craft tests for this new API from scratch.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (Mamaduka) [inline packages/block-editor/src/components/inserter/search-results.js]:
```suggestion
					blockListSettings?.prioritizedInserterBlocks || EMPTY_ARRAY,
```

It's preferred to return stable references.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (Mamaduka) [inline packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:84]: Note: We'll have to remove this hack in the futere if we want to support dynamic `allowedBlocks` and `prioritizedInserterBlocks`. See https://github.com/WordPress/gutenberg/issues/14515#issuecomment-1429218762.

**Comment by MaggieCabrera** [comment]
> A good next step here would be to identify whether there are any tests covering the `allowedBlocks` feature? If not then we will need to craft tests for this new API from scratch.

not the `allowedBlocks` specifically, but the correct order of blocks in the inserter is covered by some navigation tests already [here](https://github.com/WordPress/gutenberg/blob/f4da24145ec853cbb94f8f5800f37c13f725bd0e/test/e2e/specs/editor/blocks/navigation.spec.js#L539)

File: packages/block-editor/src/components/inner-blocks/index.native.js
Comment (getdave) [inline packages/block-editor/src/components/inner-blocks/index.native.js:82]: Do we _need_ these ones?

**Comment by getdave** [comment]
Just to flag up the follow items that are "blocking" the merge of this PR:

- native is broken and we'll need to fix that
- we need to check whether current test coverage specifically tests the new changes in this PR (doubtful)
- consider whether to [add the same functionality to the slash inserter](https://github.com/WordPress/gutenberg/pull/50510#discussion_r1191518572)
- write docs

**Comment by fluiddot** [comment]
👋 I added the label `Mobile App - Automation` to create [this Gutenberg Mobile PR automatically](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5770). The Gutenberg Mobile repo is where we run most of the E2E tests for the mobile native version, so it will tell us if these changes might have affected the app. We'll share here if all tests pass.

**Review by MaggieCabrera — DISMISSED** [review]
I thought I could dismiss my review but I can't, so let's hold off on merging until we address @getdave [comments](https://github.com/WordPress/gutenberg/pull/50510#issuecomment-1545453335)

File: packages/block-editor/src/components/inner-blocks/README.md
Comment (MaggieCabrera) [inline packages/block-editor/src/components/inner-blocks/README.md:184]: do we need to document this somewhere else or is this enough?

Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/README.md:184]: That's the official docs so I guess that's enough.

File: packages/e2e-test-utils/src/get-all-block-inserter-item-titles.js
Comment (getdave) [inline packages/e2e-test-utils/src/get-all-block-inserter-item-titles.js:23]: I changed this in https://github.com/WordPress/gutenberg/pull/50510/commits/8cb75c86b57bb06360fc279cf21897c18053bd53.

Because it does sorting on the results the tests cannot assert on the priority of the blocks.

This helper is "doing it wrong". Instead it should return raw results and the _consumer_ should do the sorting.

Let's see which tests break before deciding how to proceed from here.

**Comment by fluiddot** [comment]
> 👋 I added the label `Mobile App - Automation` to create [this Gutenberg Mobile PR automatically](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5770). The Gutenberg Mobile repo is where we run most of the E2E tests for the mobile native version, so it will tell us if these changes might have affected the app. We'll share here if all tests pass.

All E2E tests passed in the [Gutenberg Mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5770) ([CI job reference](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/20287/workflows/27266789-e147-4e5a-9a77-5b1405b997ca)) 🎊 . Note that the PR gets automatically updated when a new commit is pushed in this PR, so the tests will be re-triggered with next commits.

File: packages/block-editor/src/components/inner-blocks/index.native.js
Comment (draganescu) [reply packages/block-editor/src/components/inner-blocks/index.native.js:82]: I don't understand this question.

File: packages/block-editor/src/components/inner-blocks/index.native.js
Comment (draganescu) [inline packages/block-editor/src/components/inner-blocks/index.native.js:113]: Why is this PR adding all these other settings and not only the core feature `prioritizedInserterBlocks`?

File: packages/block-editor/src/components/inserter/index.js
Comment (draganescu) [reply packages/block-editor/src/components/inserter/index.js:433]: Wouldn't `orderInitialBlockItems` be the default order for when block's don't prioritize or for the generic canvas inserter in an empty paragraph?

Edit: no, that's for `orderInitialBlockItems` in the search results ro handle. This `orderInitialBlockItems` prop can be removed because it's effectively replaced by a public API doing the same thing.

File: packages/block-editor/src/components/inner-blocks/index.native.js
Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/index.native.js:113]: Because not doing that broke the Mobile implementation (cc @fluiddot).

File: packages/e2e-test-utils/src/get-all-block-inserter-item-titles.js
Comment (getdave) [reply packages/e2e-test-utils/src/get-all-block-inserter-item-titles.js:23]: Do we think this should be it's own sub PR?

File: packages/block-editor/src/components/inserter/index.js
Comment (draganescu) [reply packages/block-editor/src/components/inserter/index.js:427]: I would keep the private wrapper. This shape allows for easy future private APIs without all the boilerplate being re-added.

File: packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js
Comment (draganescu) [inline packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js]: What does this comment mean?

File: packages/block-editor/src/components/inner-blocks/index.native.js
Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/index.native.js:82]: The question is "why are we adding all these extra props when this PR is about `prioritizedInserterBlocks`"?

File: packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js
Comment (getdave) [reply packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js]: It was indicating that without a placeholder within the innerblocks it can be difficult to select the block using e2e tests, especially using Puppeteer. Open to suggestions for improving the comment 🙇

**Review by draganescu — APPROVED** [review]
I think this is in great shape. Left a few comments but nothing blocking 👏🏻

File: packages/block-editor/src/components/inner-blocks/index.native.js
Comment (draganescu) [reply packages/block-editor/src/components/inner-blocks/index.native.js:113]: Ah cool!

File: packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js
Comment (draganescu) [reply packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js]: > without a placeholder within the innerblocks it can be difficult to select the block using e2e tests, especially using Puppeteer, so we use an image block which has a placeholder.

Comment (getdave) [inline packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js]:
```suggestion
    // without a placeholder within the inner blocks it can be difficult to select the block using e2e tests
    // especially using Puppeteer, so we use an image block which has a placeholder.
```

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (ntsekouras) [inline packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]: We should be a bit more specific how a block variation is expressed in this API, and this applies to the docs above. We use a convention of `{blockName}/{variationName}` due to the inserter [items structure](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/selectors.js#L1816)

Comment (scruffian) [inline packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js]:
```suggestion
 * @param {string[]}             prioritizedInserterBlocks  Block names and/or block variations to be prioritized in the inserter, in the format {blockName}/{variationName}.
```

File: packages/block-editor/src/components/inner-blocks/README.md
Comment (scruffian) [inline packages/block-editor/src/components/inner-blocks/README.md]:
```suggestion
-   **Default:** - `undefined`. Determines which block types should be shown in the block inserter. For example, when inserting a block within the Navigation block we specify `core/navigation-link` and `core/navigation-link/page` as these are the most commonly used inner blocks. `prioritizedInserterBlocks` takes an array of the form {blockName}/{variationName}, where {variationName} is optional.
```

**Comment by scruffian** [comment]
I updated the docs. I say lets merge this and make sure to follow up with the slash inserter fix ASAP.

File: packages/block-editor/src/components/inserter/search-results.js
Comment (scruffian) [reply packages/block-editor/src/components/inserter/search-results.js]: Created an issue here: https://github.com/WordPress/gutenberg/issues/50635

File: packages/e2e-tests/specs/editor/plugins/inner-blocks-prioritized-inserter-blocks.test.js
Comment (ntsekouras) [inline packages/e2e-tests/specs/editor/plugins/inner-blocks-prioritized-inserter-blocks.test.js:48]: I noticed this test in the follow up PR and it doesn't seem to add any value. We should either test something about the existence/non-existence of some blocks or remove it.

Comment (getdave) [reply packages/e2e-tests/specs/editor/plugins/inner-blocks-prioritized-inserter-blocks.test.js:48]: Yes I agree. Seems pointless. It was probably me copying boilerplate code about.

We should probably change it to test what happens if you pass an empty value or empty array as the prioritized blocks argument.

File: packages/block-editor/src/components/inserter/index.js
Comment (Mamaduka) [reply packages/block-editor/src/components/inserter/index.js:427]: I was looking for why we haven't removed the private inserter. IMO, this doesn't seem like a good reason to keep things around just in case we might need them in the future.

**Pull request merged**
